### ISSUE TITLE
Error Enumerating VPC resource

### ISSUE THREAD

**Issue by zleswomp** [puppetlabs/puppetlabs-aws#260]
```
% puppet resource --debug --trace ec2_vpc
Debug: Runtime environment: puppet_version=4.3.1, ruby_version=2.1.2, run_mode=user, default_encoding=UTF-8
Debug: Evicting cache entry for environment 'production'
Debug: Caching environment 'production' (ttl = 0 sec)
Debug: Evicting cache entry for environment 'production'
Debug: Caching environment 'production' (ttl = 0 sec)
Error: Could not run: Puppet detected a problem with the information returned from AWS
when looking up ec2_vpc in us-west-2. The specific error was:

The dhcpOption ID 'dopt-0af21b6f' does not exist

Rather than report on ec2_vpc resources in an inconsistent state we have exited.
This could be because some other process is modifying AWS at the same time.
/Users/zach/.puppetlabs/etc/code/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:21:in `rescue in block in instances'
/Users/zach/.puppetlabs/etc/code/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:12:in `block in instances'
/Users/zach/.puppetlabs/etc/code/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:11:in `collect'
/Users/zach/.puppetlabs/etc/code/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:11:in `instances'
/Users/zach/src/puppet/lib/puppet/type.rb:1162:in `block in instances'
/Users/zach/src/puppet/lib/puppet/type.rb:1155:in `collect'
/Users/zach/src/puppet/lib/puppet/type.rb:1155:in `instances'
/Users/zach/src/puppet/lib/puppet/indirector/resource/ral.rb:24:in `search'
/Users/zach/src/puppet/lib/puppet/indirector/indirection.rb:265:in `search'
/Users/zach/src/puppet/lib/puppet/application/resource.rb:225:in `find_or_save_resources'
/Users/zach/src/puppet/lib/puppet/application/resource.rb:140:in `main'
/Users/zach/src/puppet/lib/puppet/application.rb:352:in `run_command'
/Users/zach/src/puppet/lib/puppet/application.rb:344:in `block in run'
/Users/zach/src/puppet/lib/puppet/util.rb:446:in `exit_on_fail'
/Users/zach/src/puppet/lib/puppet/application.rb:344:in `run'
/Users/zach/src/puppet/lib/puppet/util/command_line.rb:128:in `run'
/Users/zach/src/puppet/lib/puppet/util/command_line.rb:72:in `execute'
/Users/zach/src/puppet/bin/puppet:5:in `<main>'
```

Running `aws ec2 describe-dhcp-options` does indeed show that `"DhcpOptionsId": "dopt-0af21b6f"` is configured.

**Comment by DoctahPopp871** [comment]
While the dhcpoption exists, is it actually associated to any of your VPCs in west-2  ?

**Comment by zleswomp** [comment]
Running `aws ec2 describe-vpcs` says yes.

**Comment by DoctahPopp871** [comment]
I was just able to replicate this. 

# puppet resource ec2_vpc
Error: Could not run: Puppet detected a problem with the information returned from AWS
when looking up ec2_vpc in us-west-2. The specific error was:

The dhcpOption ID 'default' does not exist

Rather than report on ec2_vpc resources in an inconsistent state we have exited.
This could be because some other process is modifying AWS at the same time.

There are two VPC's in this region, that have a "default" dhcp option set associated to them, however the dhcpoption set does not exist, when checking the console.

**Comment by ngeorgieff** [comment]
I have the same problem:

[aws-provisioner@ip-10-100-1-238 aws]$ puppet resource --debug --trace ec2_vpc
Debug: Runtime environment: puppet_version=3.8.4, ruby_version=2.0.0, run_mode=user, default_encoding=UTF-8
Debug: Evicting cache entry for environment 'production'
Debug: Caching environment 'production' (ttl = 0 sec)
Debug: Evicting cache entry for environment 'production'
Debug: Caching environment 'production' (ttl = 0 sec)
Error: Could not run: Puppet detected a problem with the information returned from AWS
when looking up ec2_vpc in us-west-2. The specific error was:

The dhcpOption ID 'dopt-47fb0b22' does not exist

Rather than report on ec2_vpc resources in an inconsistent state we have exited.
This could be because some other process is modifying AWS at the same time.
/home/aws-provisioner/.puppet/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:21:in `rescue in block in instances'
/home/aws-provisioner/.puppet/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:12:in `block in instances'
/home/aws-provisioner/.puppet/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:11:in `collect'
/home/aws-provisioner/.puppet/modules/aws/lib/puppet/provider/ec2_vpc/v2.rb:11:in `instances'
/usr/share/ruby/vendor_ruby/puppet/type.rb:1148:in `block in instances'
/usr/share/ruby/vendor_ruby/puppet/type.rb:1141:in `collect'
/usr/share/ruby/vendor_ruby/puppet/type.rb:1141:in `instances'
/usr/share/ruby/vendor_ruby/puppet/indirector/resource/ral.rb:25:in `search'
/usr/share/ruby/vendor_ruby/puppet/indirector/indirection.rb:267:in `search'
/usr/share/ruby/vendor_ruby/puppet/application/resource.rb:225:in `find_or_save_resources'
/usr/share/ruby/vendor_ruby/puppet/application/resource.rb:142:in `main'
/usr/share/ruby/vendor_ruby/puppet/application.rb:389:in `run_command'
/usr/share/ruby/vendor_ruby/puppet/application.rb:381:in `block (2 levels) in run'
/usr/share/ruby/vendor_ruby/puppet/application.rb:507:in `plugin_hook'
/usr/share/ruby/vendor_ruby/puppet/application.rb:381:in `block in run'
/usr/share/ruby/vendor_ruby/puppet/util.rb:496:in `exit_on_fail'
/usr/share/ruby/vendor_ruby/puppet/application.rb:381:in `run'
/usr/share/ruby/vendor_ruby/puppet/util/command_line.rb:146:in `run'
/usr/share/ruby/vendor_ruby/puppet/util/command_line.rb:92:in `execute'
/usr/bin/puppet:8:in `<main>'

**Comment by diranged** [comment]
We're running into this exact same problem and its a blocker for us using this module. We have VPCs in several different regions, but whenever the `puppetlabs-aws` provider tries to enumerate all of the VPC objects, it fails with:

### Simple enumeration across all regions fails for ec2_vpc resources
```
$ puppet resource ec2_vpc

Error: Could not run: Puppet detected a problem with the information returned from AWS
when looking up ec2_vpc in us-east-1. The specific error was:

The dhcpOption ID 'dopt-16819a74' does not exist

Rather than report on ec2_vpc resources in an inconsistent state we have exited.
This could be because some other process is modifying AWS at the same time.
```

### Enumerating resources in a single region
However, if we run the same command explicitly in any one VPC, it works fine. *Including the region with the failed dhcpOption ID above*:

```
$ AWS_REGION=us-east-1 puppet resource ec2_vpc
ec2_vpc { 'DEV-OPS3':
  ensure           => 'present',
  cidr_block       => '10.48.80.0/20',
  instance_tenancy => 'default',
  region           => 'us-east-1',
}
```

### Logging statements added to the code
If I add in a bunch of log statements, I can see the failure happens on this call: https://github.com/puppetlabs/puppetlabs-aws/blob/master/lib/puppet_x/puppetlabs/aws.rb#L300

```
diff --git a/lib/puppet/provider/ec2_vpc/v2.rb b/lib/puppet/provider/ec2_vpc/v2.rb
index 4bf9b04..47357f6 100644
--- a/lib/puppet/provider/ec2_vpc/v2.rb
+++ b/lib/puppet/provider/ec2_vpc/v2.rb
@@ -35,6 +35,7 @@ Puppet::Type.type(:ec2_vpc).provide(:v2, :parent => PuppetX::Puppetlabs::Aws) do
 
   def self.vpc_to_hash(region, vpc)
     name = name_from_tag(vpc)
+    Puppet.info("vpc_to_hash(#{region}), #{vpc})")
     return {} unless name
     {
       name: name,
diff --git a/lib/puppet_x/puppetlabs/aws.rb b/lib/puppet_x/puppetlabs/aws.rb
index 18939ae..9c394d0 100644
--- a/lib/puppet_x/puppetlabs/aws.rb
+++ b/lib/puppet_x/puppetlabs/aws.rb
@@ -110,6 +110,7 @@ This could be because some other process is modifying AWS at the same time."""
         else
           config[:region] = region
         end
+        Puppet.info("config: #{config}")
         config
       end
 
@@ -295,14 +296,21 @@ This could be because some other process is modifying AWS at the same time."""
 
       def self.options_name_from_id(region, options_id)
         ec2 = ec2_client(region)
+        Puppet.info("options_name_from_id(#{region}, #{options_id})")
+        Puppet.info("options_name_from_id: starting dhcp_options=#{@dhcp_options}")
+
         @dhcp_options ||= Hash.new do |h, key|
           h[key] = unless key.nil? || key.empty?
+            Puppet.info("options_name_from_id: inner loop h=#{h} key=#{key}")
             response = ec2.describe_dhcp_options(dhcp_options_ids: [key])
+            Puppet.info("options_name_from_id: got response=#{response}")
             name_from_tag(response.dhcp_options.first)
           else
+            Puppet.info("huh, got to nil")
             nil
           end
         end
+        Puppet.info("ending dhcp_options[options_id]=#{@dhcp_options[options_id]}")
         @dhcp_options[options_id]
       end 
```

### Full output in debug mode
Here's the output in full debug mode.. note, some private info has been stripped.
```
Matts-MacBook:puppet-infrastructure diranged$ puppet resource --verbose --debug --modulepath modules/vendor:modules/specific ec2_vpc
Debug: Runtime environment: puppet_version=3.7.3, ruby_version=1.9.3, run_mode=user, default_encoding=UTF-8
Info: config: {:logger=>nil, :region=>"eu-west-1"}
Info: config: {:logger=>nil, :region=>"eu-west-1"}
Info: vpc_to_hash(eu-west-1), #<struct Aws::EC2::Types::Vpc vpc_id="vpc-a6a5cac3", state="available", cidr_block="10.64.32.0/20", dhcp_options_id="dopt-90ca22f5", tags=[...], instance_tenancy="default", is_default=false>)
Info: config: {:logger=>nil, :region=>"eu-west-1"}
Info: options_name_from_id(eu-west-1, dopt-90ca22f5)
Info: options_name_from_id: starting dhcp_options=
Info: options_name_from_id: inner loop h={} key=dopt-90ca22f5
Info: options_name_from_id: got response=#<Seahorse::Client::Response:0x007fcd03d54c88>
Info: ending dhcp_options[options_id]=
Info: config: {:logger=>nil, :region=>"ap-southeast-1"}
Info: config: {:logger=>nil, :region=>"ap-southeast-2"}
Info: config: {:logger=>nil, :region=>"eu-central-1"}
Info: vpc_to_hash(eu-central-1), #<struct Aws::EC2::Types::Vpc vpc_id="vpc-a4a345cd", state="available", cidr_block="172.31.0.0/16", dhcp_options_id="dopt-87f91dee", tags=[], instance_tenancy="default", is_default=true>)
Info: config: {:logger=>nil, :region=>"ap-northeast-2"}
Info: vpc_to_hash(ap-northeast-2), #<struct Aws::EC2::Types::Vpc vpc_id="vpc-c551b7ac", state="available", cidr_block="172.31.0.0/16", dhcp_options_id="dopt-17a9487e", tags=[], instance_tenancy="default", is_default=true>)
Info: config: {:logger=>nil, :region=>"ap-northeast-1"}
Info: config: {:logger=>nil, :region=>"us-east-1"}
Info: vpc_to_hash(us-east-1), #<struct Aws::EC2::Types::Vpc vpc_id="vpc-a86710cd", state="available", cidr_block="10.48.80.0/20", dhcp_options_id="dopt-16819a74", tags=[...], instance_tenancy="default", is_default=false>)
Info: config: {:logger=>nil, :region=>"us-east-1"}
Info: options_name_from_id(us-east-1, dopt-16819a74)
Info: options_name_from_id: starting dhcp_options={"dopt-90ca22f5"=>nil}
Info: options_name_from_id: inner loop h={"dopt-90ca22f5"=>nil} key=dopt-16819a74
Error: Could not run: Puppet detected a problem with the information returned from AWS
when looking up ec2_vpc in us-east-1. The specific error was:

The dhcpOption ID 'dopt-16819a74' does not exist

Rather than report on ec2_vpc resources in an inconsistent state we have exited.
This could be because some other process is modifying AWS at the same time.
Matts-MacBook:puppet-infrastructure diranged$ 
```

###  Finally, doing it with the API...
Just for fun, I loaded up the Ruby shell and I did the search by hand:
```
1.9.3-p551 :066 > ec2 = Aws::EC2::Client.new(:region => 'us-east-1')
 => #<Aws::EC2::Client> 
1.9.3-p551 :067 > ec2.describe_dhcp_options(dhcp_options_ids: ['dopt-16819a74'])
 => #<struct Aws::EC2::Types::DescribeDhcpOptionsResult dhcp_options=[#<struct Aws::EC2::Types::DhcpOptions dhcp_options_id="dopt-16819a74", dhcp_configurations=[#<struct Aws::EC2::Types::DhcpConfiguration key="domain-name", values=[#<struct Aws::EC2::Types::AttributeValue value="ec2.internal">]>, #<struct Aws::EC2::Types::DhcpConfiguration key="domain-name-servers", values=[#<struct Aws::EC2::Types::AttributeValue value="AmazonProvidedDNS">]>], tags=[#<struct Aws::EC2::Types::Tag key="Name", value="">]>]> 
1.9.3-p551 :068 > 
```

**Comment by diranged** [comment]
@kenazk, any chance you can take a look at this issue?

**Comment by diranged** [comment]
Ping???

**Comment by diranged** [comment]
Ping! :)

**Comment by kenazk** [comment]
@diranged Sorry for the delay. This issue is in our next sprint.

**Comment by DavidS** [comment]
Hi @diranged ,

I think this was addressed in https://github.com/puppetlabs/puppetlabs-aws/commit/612e8b0c50628b277dbb9464a1c96786446384af which is already released as 1.4.0. It'd be great if you could verify that fix.

Cheers

**Comment by diranged** [comment]
It does indeed seem fixed in 1.4.0... I've just done a little testing so far, but it no longer crashes. I'll do more testing this week to confirm we see no other odd behaviors.

**Comment by DavidS** [comment]
Excellent! Please feel free to open any new issues you uncover!
